sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoApiClient.kt

```diff
@@ -68,55 +71,81 @@ internal object KlaviyoApiClient : ApiClient {
             Registry.dataStore.store(request.uuid, request.toJson())
         }
 
-        Registry.dataStore.store(
-            QUEUE_KEY,
-            JSONArray(apiQueue.map { it.uuid }).toString()
-        )
+        persistQueue()
     }
 
     /**
-     * Awaken queued requests from persistent store
+     * Gets the current size of the API queue
+     *
+     * @return number of requests in the batch queue
+     */
+    fun getQueueSize(): Int = apiQueue.size
+
+    /**
+     * Reset the in-memory queue to the queue from data store
      */
     fun restoreQueue() {
-        // TODO JSON type safety checks
+        while (apiQueue.isNotEmpty()) {
+            apiQueue.remove()
+        }
+
         Registry.dataStore.fetch(QUEUE_KEY)?.let {
-            val queue = JSONArray(it)
-            Array(queue.length()) { i -> queue.optString(i) }
-        }?.map { uuid ->
+            try {
+                val queue = JSONArray(it)
+                Array(queue.length()) { i -> queue.optString(i) }
+            } catch (exception: JSONException) {
+                emptyArray<String>()
+            }
+        }?.forEach { uuid ->
             Registry.dataStore.fetch(uuid)?.let { json ->
-                val jsonObject = JSONObject(json)
-                val request = KlaviyoApiRequest.fromJson(jsonObject)
-                enqueueRequest(request)
+                try {
+                    val request = KlaviyoApiRequest.fromJson(JSONObject(json))
+                    if (!apiQueue.contains(request)) {
+                        apiQueue.offer(request)
+                    }
+                } catch (exception: JSONException) {
+                    Registry.dataStore.clear(uuid)
+                }
             }
         }
+
+        if (apiQueue.isNotEmpty()) initBatch()
+
+        persistQueue()
```

Comment:
is it necessary to write to disk right after reading from disk?

Reply:
I figured it just persists any mutations that had to be made in case of a JSON parse error above. But I guess it should be conditional, only do this if we know something was changed, otherwise it is a waste.